### PR TITLE
Port to clap 4, drop custom derive

### PR DESCRIPTION
**Pull request by elprans**
Move the CLI over to clap-4.4.6, drop `EdbClap` and use clap's
derive instead.

Custom derive allowed is to implement various hacks and logic in
argument propagation and annotation, but the cost of maintaining what is
essentially an old fork of `clap::_derive` is quite high and stands in
the way of easy `clap` upgrades and the ability to use its newer
features.

Fortunately, I've been able to find ways to achieve the same argument
propagation behavior with normal `global` args and a little bit of
hackery at app initialization, so instead of doing

    #[edb(inherit(CloudOptions))]
    command or subcommand

you should now do

    #[command(flatten)]
    pub cloud_opts: CloudOptions

in the parent command of where you want to begin propagation.  Though
this is slightly more verbose (the current derive is a bit more magical
about propagation), this approach has the benefit of making the options
available closer to the actual subcommand as opposed to passing the
root `Options` all the way through, but I've left that cleanup for
later.

### CODE REVIEW

File: src/options.rs [hunk 1]
```diff
@@ -491,31 +532,27 @@ fn update_main_help(mut app: clap::Command) -> clap::Command {
...
-    let subcmd_index = help.find("SUBCOMMANDS:").unwrap();
-    let mut help = help[..subcmd_index].replacen("edgedb", "EdgeDB CLI", 1);
-    help.push_str(&sub_cmd);
+    let help = format!("{}", app.render_help().ansi()).to_string();
+    let subcmd_index = help.find("Commands:").unwrap();
+    let opt_index = help.find("Options:").unwrap();
 
-    help = help.replacen(
-        CONN_OPTIONS_GROUP,
-        &markdown::format_markdown(CONN_OPTIONS_GROUP).trim(),
```
Comment (fantix) [inline src/options.rs:506]: `markdown::format_markdown` is no longer used.

**Review by fantix — APPROVED** [review]
The backslash commands in REPL are crashing in my manual test. Otherwise this looks good and works fine (with a basic manual test).

```
edgedb> \l
thread 'tokio-runtime-worker' panicked at 'Command edgedb-cli: command name `help` is duplicated', /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/debug_asserts.rs:334:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f4956053816439a5884cb2ad1247835858f92218/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/f4956053816439a5884cb2ad1247835858f92218/library/core/src/panicking.rs:67:14
   2: clap_builder::builder::debug_asserts::assert_app
   3: clap_builder::builder::command::Command::_build_self
             at /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/command.rs:3985:13
   4: clap_builder::builder::command::Command::_do_parse
             at /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/command.rs:3856:9
   5: clap_builder::builder::command::Command::try_get_matches_from_mut
             at /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/command.rs:791:9
   6: clap_builder::builder::command::Command::try_get_matches_from
             at /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/command.rs:709:9
   7: edgedb::commands::backslash::parse
             at ./src/commands/backslash.rs:439:5
```

**Comment by Dhghomon** [comment]
> The backslash commands in REPL are crashing in my manual test. Otherwise this looks good and works fine (with a basic manual test).
> 
> ```
> edgedb> \l
> thread 'tokio-runtime-worker' panicked at 'Command edgedb-cli: command name `help` is duplicated', /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/debug_asserts.rs:334:9
> stack backtrace:
>    0: rust_begin_unwind
>              at /rustc/f4956053816439a5884cb2ad1247835858f92218/library/std/src/panicking.rs:578:5
>    1: core::panicking::panic_fmt
>              at /rustc/f4956053816439a5884cb2ad1247835858f92218/library/core/src/panicking.rs:67:14
>    2: clap_builder::builder::debug_asserts::assert_app
>    3: clap_builder::builder::command::Command::_build_self
>              at /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/command.rs:3985:13
>    4: clap_builder::builder::command::Command::_do_parse
>              at /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/command.rs:3856:9
>    5: clap_builder::builder::command::Command::try_get_matches_from_mut
>              at /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/command.rs:791:9
>    6: clap_builder::builder::command::Command::try_get_matches_from
>              at /home/fantix/.cargo/registry/src/index.crates.io-6f17d22bba15001f/clap_builder-4.4.6/src/builder/command.rs:709:9
>    7: edgedb::commands::backslash::parse
>              at ./src/commands/backslash.rs:439:5
> ```

Same when using it on my end. So far have only noticed this as an outright bug, while the necessary extra `--` when passing in an option might confuse users so thinking of a good place to note that in the menu (e.g. now `edgedb -- --help-connect` instead of `edgedb --help-connect`.

Clap doing away with colour is a bit unfortunate but otherwise it's a huge step up and I love being able to do away with the custom derive.

That's what I've noticed so far.

File: src/options.rs [hunk 2]
```diff
@@ -32,109 +30,112 @@ use crate::repl::OutputFormat;
...
-pub mod describe;
-
-const MAX_TERM_WIDTH: usize = 90;
+const MAX_TERM_WIDTH: usize = 100;
 const MIN_TERM_WIDTH: usize = 50;
 
 const CONN_OPTIONS_GROUP: &str =
-    "CONNECTION OPTIONS (`edgedb --help-connect` to see full list)";
-const CLOUD_OPTIONS_GROUP: &str = "CLOUD OPTIONS";
+    "Connection Options (edgedb --help-connect to see full list)";
```
Comment (Dhghomon) [inline src/options.rs:37]: Now is `edgedb -- --help-connect` to see the full list

Comment (elprans) [reply src/options.rs:37]: Why the double dash? `edgedb --help-connect` works just fine as before.  Am I missing something?

Comment (Dhghomon) [reply src/options.rs:37]: Sorry, that was when running the executable through cargo which I usually do. Please ignore

**Review by Dhghomon — APPROVED** [review]
Okay, spent a good while with the code and trying out the CLI commands and have only seen the same issue that @fantix noticed with the duplicate command name, everything else looks good and matches what clap recommends too. I'll keep an eye on further changes and let you know if I spot anything unusual.

**Pull request merged**